File: packages/ironfish-native-module/ios/IronfishNativeModule.swift
import ExpoModulesCore

struct ExpoKey: Record {
  @Field
  var spendingKey: String

  @Field
  var viewKey: String

  @Field
  var incomingViewKey: String

  @Field
  var outgoingViewKey: String

  @Field
  var publicAddress: String

  @Field
  var proofAuthorizingKey: String
}

struct ExpoOutput : Record {
  @Field
  var index: UInt32 = 0

  @Field
  var note: String
}

struct SpendComponentsInput: Record {
  @Field
  var components: [SpendComponentInput]
}

struct SpendComponentInput: Record {
  @Field
  var note: String

  @Field
  var witnessRootHash: String

  @Field
  var witnessTreeSize: String

  @Field
  var witnessAuthPath: [WitnessNodeInput]
}

struct WitnessNodeInput: Record {
  @Field
  var side: String

  @Field
  var hashOfSibling: String
}

struct OutputsInput: Record {
  @Field
  var outputs: [String]
}

public class IronfishNativeModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('IronfishNativeModule')` in JavaScript.
    Name("IronfishNativeModule")

    Function("generateKey") { () -> ExpoKey in
      let k = generateKey()

      return ExpoKey(
        spendingKey: Field(wrappedValue: k.spendingKey),
        viewKey: Field(wrappedValue: k.viewKey),
        incomingViewKey: Field(wrappedValue: k.incomingViewKey),
        outgoingViewKey: Field(wrappedValue: k.outgoingViewKey),
        publicAddress: Field(wrappedValue: k.publicAddress),
        proofAuthorizingKey: Field(wrappedValue: k.proofAuthorizingKey)
      )
    }

    Function("spendingKeyToWords") { (privateKey: String, languageCode: Int32) throws -> String in
      let phrase = try spendingKeyToWords(privateKey: privateKey, languageCode: languageCode)
      return phrase
    }

    Function("wordsToSpendingKey") { (words: String, languageCode: Int32) throws -> String? in
      guard let k = try? wordsToSpendingKey(words: words, languageCode: languageCode) else {
        return nil
      }

      return k
    }

    Function("generateKeyFromPrivateKey") { (privateKey: String) throws -> ExpoKey? in
      guard let k = try? generateKeyFromPrivateKey(privateKey: privateKey) else {
        return nil
      }

      return ExpoKey(
        spendingKey: Field(wrappedValue: k.spendingKey),
        viewKey: Field(wrappedValue: k.viewKey),
        incomingViewKey: Field(wrappedValue: k.incomingViewKey),
        outgoingViewKey: Field(wrappedValue: k.outgoingViewKey),
        publicAddress: Field(wrappedValue: k.publicAddress),
        proofAuthorizingKey: Field(wrappedValue: k.proofAuthorizingKey)
      )
    }

    Function("isValidPublicAddress") { (hexAddress: String) -> Bool in
      return isValidPublicAddress(hexAddress: hexAddress)
    }

    AsyncFunction("unpackGzip") { (gzipPath: String, outputPath: String) -> Bool in
      return unpackGzip(gzipPath: gzipPath, outputPath: outputPath)
    }

    AsyncFunction("readPartialFile") { (path: String, offset: UInt32, length: UInt32) -> Data in
      return readPartialFile(path: path, offset: offset, length: length)
    }

    AsyncFunction("decryptNotesForOwner") { (noteEncrypteds: [String], incomingHexKey: String, promise: Promise) in
      Task {
        let decryptedNotes = try await decryptNotesForOwner(noteEncrypteds: noteEncrypteds, incomingHexKey: incomingHexKey)
        let expoOutputs = decryptedNotes.map { note in
          ExpoOutput(index: note.index, note: Field(wrappedValue: note.note))
        }
        promise.resolve(expoOutputs)
      }
    }

    AsyncFunction("decryptNotesForSpender") { (noteEncrypteds: [String], outgoingHexKey: String, promise: Promise) in
      Task {
        let decryptedNotes = try await decryptNotesForSpender(noteEncrypteds: noteEncrypteds, outgoingHexKey: outgoingHexKey)
        let expoOutputs = decryptedNotes.map { note in
          ExpoOutput(index: note.index, note: Field(wrappedValue: note.note))
        }
        promise.resolve(expoOutputs)
      }
    }

    AsyncFunction("nullifier") { (note: String, position: String, viewKey: String) throws -> String in
      return try nullifier(note: note, position: position, viewKey: viewKey)
    }

    AsyncFunction ("createNote") { (owner: Data, value: String, memo: Data, assetId: Data, sender: Data) throws -> Data in
      do {
          let note = try createNote(params: NoteParams(owner: owner, value: UInt64(value)!, memo: memo, assetId: assetId, sender: sender))
          return note
      } catch let error as NSError {
          print("Unexpected error: \(error.debugDescription)")
          throw error
      }
    }

    AsyncFunction("createTransaction") { (transactionVersion: UInt8, transactionFee: String, expirationSequence: UInt32, spendComponents: SpendComponentsInput, outputs: OutputsInput, spendingKey: Data) throws  -> Data in
      let spendComponentsConverted = spendComponents.components.map { spendComponent in
          let witnessAuthPath: [WitnessNode] = spendComponent.witnessAuthPath.map { WitnessNode(side: $0.side, hashOfSibling: Data(hexString: $0.hashOfSibling)!) }
          return SpendComponents(note: Data(hexString: spendComponent.note)!, witnessRootHash: Data(hexString: spendComponent.witnessRootHash)!, witnessTreeSize: UInt64(spendComponent.witnessTreeSize)!, witnessAuthPath: witnessAuthPath)
      }
      do {
        let transaction = try createTransaction(transactionVersion: transactionVersion, transactionFee: UInt64(transactionFee)!, expirationSequence: expirationSequence, spendComponents: spendComponentsConverted, outputs: outputs.outputs.map {Data(hexString: $0)!}, spendingKey: spendingKey)
          return transaction
      } catch let error as NSError {
          print("Unexpected error: \(error.debugDescription)")
          throw error
      }
    }
  }
}

extension Data {
    init?(hexString: String) {
        let length = hexString.count / 2
        var data = Data(capacity: length)
        for i in 0..<length {
            let j = hexString.index(hexString.startIndex, offsetBy: i*2)
            let k = hexString.index(j, offsetBy: 2)
            let bytes = hexString[j..<k]
            if var num = UInt8(bytes, radix: 16) {
                data.append(&num, count: 1)
            } else {
                return nil
            }
        }
        self = data
    }
}

File: packages/ironfish-native-module/android/src/main/java/expo/modules/ironfishnativemodule/IronfishNativeModule.kt
package expo.modules.ironfishnativemodule

import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Record

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class ExpoKey(
  @Field
  val spendingKey: String,
  @Field
  val viewKey: String,
  @Field
  val incomingViewKey: String,
  @Field
  val outgoingViewKey: String,
  @Field
  val publicAddress: String,
  @Field
  val proofAuthorizingKey: String
) : Record

class ExpoOutput(
  @Field
  val index: Int,

  @Field
  val note: String,
 ) : Record

class SpendComponentsInput(
  @Field
  val components: List<SpendComponentInput>,
 ) : Record

class SpendComponentInput(
  @Field
  val note: String,

  @Field
  val witnessRootHash: String,

  @Field
  val witnessTreeSize: String,

  @Field
  val witnessAuthPath: List<WitnessNodeInput>,
) : Record

class WitnessNodeInput (
  @Field
  val side: String,

  @Field
  val hashOfSibling: String,
): Record

class OutputsInput (
  @Field
  val outputs: List<String>,
): Record

class IronfishNativeModule : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('IronfishNativeModule')` in JavaScript.
    Name("IronfishNativeModule")

    Function("generateKey") { ->
      val k = uniffi.rust_lib.generateKey()

      ExpoKey(
        k.spendingKey,
        k.viewKey,
        k.incomingViewKey,
        k.outgoingViewKey,
        k.publicAddress,
        k.proofAuthorizingKey
      )
    }

    Function("spendingKeyToWords") { privateKey: String, languageCode: Int ->
      uniffi.rust_lib.spendingKeyToWords(privateKey, languageCode)
    }

    Function("wordsToSpendingKey") { words: String, languageCode: Int ->
      uniffi.rust_lib.wordsToSpendingKey(words, languageCode)
    }

    Function("generateKeyFromPrivateKey") { privateKey: String ->
      val k = uniffi.rust_lib.generateKeyFromPrivateKey(privateKey)

      ExpoKey(
        k.spendingKey,
        k.viewKey,
        k.incomingViewKey,
        k.outgoingViewKey,
        k.publicAddress,
        k.proofAuthorizingKey
      )
    }

    Function("generatePublicAddressFromIncomingViewKey") { incomingViewKey: String ->
      uniffi.rust_lib.generatePublicAddressFromIncomingViewKey(incomingViewKey)
    }

    Function("isValidPublicAddress") { hexAddress: String ->
      uniffi.rust_lib.isValidPublicAddress(hexAddress)
    }

    AsyncFunction("unpackGzip") { gzipPath: String, outputPath: String ->
      uniffi.rust_lib.unpackGzip(gzipPath, outputPath)
    }

    AsyncFunction("readPartialFile") { path: String, offset: Long, length: Long ->
        uniffi.rust_lib.readPartialFile(path, offset.toUInt(), length.toUInt())
    }

    AsyncFunction("decryptNotesForOwner") { noteEncrypteds: List<String>, incomingHexKey: String, promise: Promise ->
      CoroutineScope(Dispatchers.Default).launch {
        try {
          val decryptedNotes = uniffi.rust_lib.decryptNotesForOwner(noteEncrypteds, incomingHexKey)
          val expoOutputs = decryptedNotes.map { note ->
              ExpoOutput(index = note.index.toInt(), note = (note.note))
          }
          promise.resolve(expoOutputs)
        } catch (e: Exception) {
          promise.reject(CodedException(e))
        }
      }
    }

    AsyncFunction("decryptNotesForSpender") { noteEncrypteds: List<String>, outgoingHexKey: String, promise: Promise ->
      CoroutineScope(Dispatchers.Default).launch {
        try {
            val decryptedNotes = uniffi.rust_lib.decryptNotesForSpender(noteEncrypteds, outgoingHexKey)
            val expoOutputs = decryptedNotes.map { note ->
                ExpoOutput(index = note.index.toInt(), note = note.note)
            }
          promise.resolve(expoOutputs)
        } catch (e: Exception) {
          promise.reject(CodedException(e))
        }
      }
    }

    AsyncFunction("nullifier") { note: String, position: String, viewKey: String ->
      uniffi.rust_lib.nullifier(note, position, viewKey)
    }

    AsyncFunction("createNote") { owner: ByteArray, value: String, memo: ByteArray, assetId: ByteArray, sender: ByteArray ->
      try {
          uniffi.rust_lib.createNote(uniffi.rust_lib.NoteParams(owner, value.toULong(), memo, assetId, sender))
      } catch (e: Exception) {
          println("Unexpected error: ${e.message}")
          throw e
      }
    }

    AsyncFunction("createTransaction") { transactionVersion: Int, transactionFee: String, expirationSequence: Int, spendComponents: SpendComponentsInput, outputs: OutputsInput, spendingKey: ByteArray ->
      val spendComponentsConverted = spendComponents.components.map { spendComponent ->
        val witnessAuthPath = spendComponent.witnessAuthPath.map { uniffi.rust_lib.WitnessNode(it.side, it.hashOfSibling.hexStringToByteArray()) }
        uniffi.rust_lib.SpendComponents(spendComponent.note.hexStringToByteArray(), spendComponent.witnessRootHash.hexStringToByteArray(), spendComponent.witnessTreeSize.toULong(), witnessAuthPath)
      }
      try {
        val transaction = uniffi.rust_lib.createTransaction(transactionVersion.toUByte(), transactionFee.toULong(), expirationSequence.toUInt(), spendComponentsConverted, outputs.outputs.map { it.hexStringToByteArray() }, spendingKey)
        transaction
      } catch (e: Exception) {
        println("Unexpected error: ${e.message}")
        throw e
      }
    }

    AsyncFunction("hashTransaction") { transaction: ByteArray ->
      uniffi.rust_lib.hashTransaction(transaction)
    }
  }
}

fun String.hexStringToByteArray(): ByteArray {
    val len = this.length
    require(len % 2 == 0) { "Hex string must have an even length" }

    return ByteArray(len / 2) { i ->
        val index = i * 2
        ((this[index].digitToInt(16) shl 4) + this[index + 1].digitToInt(16)).toByte()
    }
}
